src/node/stage/approved.rs

```diff
@@ -677,225 +735,268 @@ impl Approved {
+                self.handle_neighbour_info_event(core, elders_info)?
             }
             AccumulatingEvent::TheirKeyInfo(key_info) => {
-                self.handle_their_key_info_event(key_info)?
-            }
-            AccumulatingEvent::AckMessage(_payload) => {
-                // Update their_knowledge is handled within the chain.
+                self.handle_their_key_info_event(core, key_info)?
             }
+            AccumulatingEvent::AckMessage(payload) => self.handle_ack_message_event(payload),
             AccumulatingEvent::SendAckMessage(payload) => {
                 self.handle_send_ack_message_event(core, payload)?
             }
             AccumulatingEvent::ParsecPrune => self.handle_prune_event(core)?,
             AccumulatingEvent::Relocate(payload) => self.handle_relocate_event(core, payload)?,
             AccumulatingEvent::RelocatePrepare(pub_id, count) => {
-                self.handle_relocate_prepare_event(pub_id, count);
+                self.handle_relocate_prepare_event(core, pub_id, count);
             }
             AccumulatingEvent::User(payload) => self.handle_user_event(core, payload)?,
         }
 
         Ok(())
     }
 
-    fn handle_relocate_polled(&mut self, details: RelocateDetails) -> Result<(), RoutingError> {
-        if !self.chain.is_self_elder() {
-            return Ok(());
-        }
-
-        self.vote_for_relocate_prepare(details, INITIAL_RELOCATE_COOL_DOWN_COUNT_DOWN);
-
-        Ok(())
-    }
-
-    fn handle_promote_and_demote_elders(
+    // Handles an accumulated parsec Observation for genesis.
+    //
+    // The related_info is the serialized shared state that will be the starting
+    // point when processing parsec data.
+    fn handle_genesis_event(
         &mut self,
-        new_infos: Vec<EldersInfo>,
-    ) -> Result<(), RoutingError> {
-        if !self.chain.is_self_elder() {
-            return Ok(());
-        }
+        _group: &BTreeSet<PublicId>,
+        related_info: &[u8],
+    ) -> Result<()> {
+        // `related_info` is empty only if this is the `first` node.
+        let new_state = if !related_info.is_empty() {
+            Some(bincode::deserialize(related_info)?)
+        } else {
+            None
+        };
 
-        for info in new_infos {
-            let participants: BTreeSet<_> = info.member_ids().copied().collect();
-            let _ = self.dkg_cache.insert(participants.clone(), info);
-            self.vote_for_event(AccumulatingEvent::StartDkg(participants));
-        }
+        // On split membership may need to be checked again.
+        self.members_changed = true;
+        self.shared_state.update(new_state);
 
         Ok(())
     }
 
     fn handle_online_event(&mut self, core: &mut Core, payload: OnlinePayload) -> Result<()> {
-        if !self.chain.can_add_member(payload.p2p_node.public_id()) {
-            info!("ignore Online: {:?}.", payload);
-        } else {
+        assert!(!self.churn_in_progress);
```

Comment:
I think we shall not have an assertion within the production code?
same comment to other places

Reply:
The assertion existed before, but yeah, it's controversial. I'll remove it.

Reply:
OK. Will replace it with log and early exit.